### PR TITLE
Asynchrounous multisig program

### LINKED ISSUE #348: Asynchronous multisig example
Same as #259, but with arbitrary number of signing parties involved

### PR DESCRIPTION
**Pull request by gshep**
Closes #348 #395  .

### CODE REVIEW

File: examples/async-multisig/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+fn hex_to_id(hex: &str) -> Result<ProgramId, ()> {
+    let hex = hex.strip_prefix("0x").unwrap_or(hex);
+
+    hex::decode(hex)
+        .map(|bytes| ProgramId::from_slice(&bytes))
+        .map_err(|_| ())
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
```
Comment (NikVolf) [inline examples/async-multisig/src/lib.rs:49]: Can we make it strongly-typed?

Vec<ProgramId> would be cool

File: examples/async-multisig/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    let mut requests: Vec<_> = unsafe { &SIGNATORIES }
+        .iter()
+        .enumerate()
+        .map(|(i, s)| {
+            msg_async::send_and_wait_for_reply(*s, &encoded, GAS_LIMIT, 0).map(move |r| (i, r))
+        })
+        .collect();
+
+    let mut threshold = 0usize;
+    while !requests.is_empty() {
```
Comment (NikVolf) [inline examples/async-multisig/src/lib.rs:90]: This will send messages one-by-one

We need to send them all at once

Comment (gshep) [reply examples/async-multisig/src/lib.rs:90]: The following setlect_all line seems to send messages simultaneously

File: examples/async-multisig/src/lib.rs (hunk 1 above)
Comment (gshep) [reply examples/async-multisig/src/lib.rs:49]: Do you mean something like this:
```rust
#[derive(Debug, Decode, TypeInfo)]
pub struct InputArgs {
     pub threshold: usize,
     pub destination: ProgramId,
     pub signatories: Vec<ProgramId>,
 }
 
 gstd::metadata! {
    title: "demo async multisig",
    init:
        input: InputArgs,
 }
 ```
 ?

Comment (NikVolf) [reply examples/async-multisig/src/lib.rs:49]: Aye

File: examples/async-multisig/src/lib.rs (hunk 2 above)
Comment (NikVolf) [reply examples/async-multisig/src/lib.rs:90]: oh nice trick, I couldn't even read it at first

seems ok, but will require quite a good amount of tests to show it's working in many situations

File: examples/async-multisig/src/lib.rs (hunk 1 above)
Comment (gshep) [reply examples/async-multisig/src/lib.rs:46]: done

File: examples/async-multisig/src/lib.rs (hunk 2 above)
Comment (gshep) [reply examples/async-multisig/src/lib.rs:81]: sure, working on it

File: examples/async-multisig/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+                        let mut signature: Signature = Default::default();
+                        if response.signature.len() == signature.0.len() {
+                            signature.as_mut().copy_from_slice(&response.signature);
+                            Some(signature)
+                        } else {
+                            None
+                        }
+                    })
+                    .ok()
+                    .flatten()
```
Comment (shamilsan) [inline examples/async-multisig/src/lib.rs:97]: Avoid excessive `flatten`.
```suggestion
                SignResponse::decode(&mut &bytes[..])
                    .ok()
                    .and_then(|response| {
                        // the same way as in verify.rs from subkey
                        let mut signature: Signature = Default::default();
                        if response.signature.len() == signature.0.len() {
                            signature.as_mut().copy_from_slice(&response.signature);
                            Some(signature)
                        } else {
                            None
                        }
                    })
```

File: examples/async-multisig/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        })
+        .collect();
+
+    let mut threshold = 0usize;
+    while !requests.is_empty() {
+        let ((i, result), _, remaining) = future::select_all(requests).await;
+
+        threshold += result
+            .ok()
+            .map(|bytes| {
```
Comment (shamilsan) [inline examples/async-multisig/src/lib.rs:84]:
```suggestion
            .and_then(|bytes| {
```

File: examples/async-multisig/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                            signature.as_mut().copy_from_slice(&response.signature);
+                            Some(signature)
+                        } else {
+                            None
+                        }
+                    })
+                    .ok()
+                    .flatten()
+            })
+            .flatten()
```
Comment (shamilsan) [inline examples/async-multisig/src/lib.rs:99]:
```suggestion
```

**Review by NikVolf — APPROVED** [review]
other some comments above -- cool

**Review by shamilsan — APPROVED** [review]

**Pull request merged**